### PR TITLE
OBG 630: Encode operation into request signing flow

### PR DESCRIPTION
**Pull request by ivaniuminov**
Init commit, tests will be adjusted after dynamic signature creation will be added
https://github.com/adorsys/open-banking-gateway/pull/639

### CODE REVIEW

File: opba-api-security/src/main/java/de/adorsys/opba/api/security/domain/OperationType.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package de.adorsys.opba.api.security.domain;
+
+public enum OperationType {
+    AIS,
+    BANK_SEARCH,
+    TOKEN
```
Comment (valb3r) [inline opba-api-security/src/main/java/de/adorsys/opba/api/security/domain/OperationType.java:6]: This is not TOKEN but rather CONFIRM_CONSENT

File: fintech-examples/fintech-impl/src/main/java/de/adorsys/opba/fintech/impl/config/FeignConfig.java [hunk 2]
```diff
@@ -49,7 +51,8 @@ private void fillSecurityHeaders(RequestTemplate requestTemplate) {
 
     private String calculateSignature(Request request, Instant instant) {
         String xRequestId = request.headers().get(X_REQUEST_ID).stream().findFirst().orElse(null);
-        DataToSign dataToSign = new DataToSign(UUID.fromString(xRequestId), instant);
+        String operationType = request.headers().get(X_OPERATION_TYPE).stream().findFirst().orElse(null);
+        DataToSign dataToSign = new DataToSign(UUID.fromString(xRequestId), instant, OperationType.valueOf(operationType));
```
Comment (valb3r) [inline fintech-examples/fintech-impl/src/main/java/de/adorsys/opba/fintech/impl/config/FeignConfig.java:55]: You need to:
1. From the request, path deduce OperationType (probably some map in application.yml)
2. Check that OperationType from request equals to OperationType you calculated from pt. 1
3. And only after compute the signature
In essence, OperationType should identify endpoint (or endpoint 'class')

File: opba-api-security/src/main/java/de/adorsys/opba/api/security/domain/DataToSign.java [hunk 3]
```diff
@@ -59,9 +61,8 @@ private String mapAdditionalFieldsToString() {
             return nodes.asText();
         }
 
-        Map<String, String> unSortedMap = objectToUnsortedMap(nodes);
-        LinkedHashMap<String, String> sortedMap = sortMap(unSortedMap);
-        LinkedHashMap<String, String> filteredMap = valueNotNullFilter(sortedMap);
+        LinkedHashMap<String, String> map = objectToMap(nodes);
```
[the remark this answers is not in the document]
Comment (ivaniuminov) [reply opba-api-security/src/main/java/de/adorsys/opba/api/security/domain/DataToSign.java:64]: Done, LinkedHashMap is used no more

File: opba-api-security/src/main/java/de/adorsys/opba/api/security/domain/OperationType.java (hunk 1 above)
Comment (ivaniuminov) [reply opba-api-security/src/main/java/de/adorsys/opba/api/security/domain/OperationType.java:6]: Renamed

File: fintech-examples/fintech-impl/src/main/java/de/adorsys/opba/fintech/impl/config/FeignConfig.java (hunk 2 above)
Comment (ivaniuminov) [reply fintech-examples/fintech-impl/src/main/java/de/adorsys/opba/fintech/impl/config/FeignConfig.java:55]: Validation filter was added

File: opba-api-security/src/main/java/de/adorsys/opba/api/security/config/OperationTypeProperties.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import javax.validation.constraints.NotEmpty;
+import java.util.concurrent.ConcurrentHashMap;
+
+@Data
+@Validated
+@Configuration
+@ConfigurationProperties(prefix = "security.endpoint")
+public class OperationTypeProperties {
+    @NotEmpty
+    private ConcurrentHashMap<String, String> allowedPath;
```
Comment (valb3r) [inline opba-api-security/src/main/java/de/adorsys/opba/api/security/config/OperationTypeProperties.java:17]: Also please add
`private ConcurrentHashMap<@NotBlank String, @NotBlank  String> allowedPath;`

Comment (valb3r) [inline opba-api-security/src/main/java/de/adorsys/opba/api/security/config/OperationTypeProperties.java:17]: It is better to use Map interface - there will be no concurrent updates of this `allowedPath` field

File: opba-api-security/src/main/java/de/adorsys/opba/api/security/filter/OperationTypeFilter.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+@Slf4j
+@Order(0)
+@Component
+@RequiredArgsConstructor
+public class OperationTypeFilter extends AbstractSecurityFilter {
```
Comment (valb3r) [inline opba-api-security/src/main/java/de/adorsys/opba/api/security/filter/OperationTypeFilter.java:20]: It is better to have one security filter - to have core security concept in one place (at least while class size is acceptable)

File: opba-api-security/src/main/java/de/adorsys/opba/api/security/config/OperationTypeProperties.java (hunk 4 above)
Comment (ivaniuminov) [reply opba-api-security/src/main/java/de/adorsys/opba/api/security/config/OperationTypeProperties.java]: added

Comment (ivaniuminov) [reply opba-api-security/src/main/java/de/adorsys/opba/api/security/config/OperationTypeProperties.java]: Map is used

File: opba-api-security/src/main/java/de/adorsys/opba/api/security/filter/OperationTypeFilter.java (hunk 5 above)
Comment (ivaniuminov) [reply opba-api-security/src/main/java/de/adorsys/opba/api/security/filter/OperationTypeFilter.java]: Second filter and abstract one were removed

**Pull request merged**